File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        Map<Integer,Integer> map = new HashMap<>();

        map.put(0,1);

        int preSum = 0;
        int count = 0;

        for(int right = 0;right <nums.length;right++ )
        {

            preSum += nums[right];

            if(map.containsKey(preSum - goal))
            {
                count += map.get(preSum - goal);
            }

            map.put(preSum , map.getOrDefault(preSum, 0) + 1);



        }
        return count;
    }
}

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.cpp
class Solution {
public:

    int atMost(vector<int>& nums, int goal)
    {
         if (goal < 0) return 0;      
        
        int right = 0;
        int left = 0;
        int sum = 0;
        int count = 0;
        while(right < nums.size())
        {

           
            sum += nums[right];
            while(sum > goal)
            {
                sum -= nums[left];
                left++;
            }
            
            count += right - left + 1;
            right++;
        }

        return count;
    }



    int numSubarraysWithSum(vector<int>& nums, int goal) {
        return atMost(nums, goal) - atMost(nums, goal - 1);
    }
};